File: libcpp/WUS_Control/bursting_state.cpp
#include "bursting_state.h"

stateName BurstingState :: get_name(void) {
    return BURSTING_STATE;
}

void BurstingState :: enter(StateController* ctrl) {
    // Serial.println("bursting");

    ctrl->set_rgbLED(0, 0, 255); // blue
    this->set_parameters(ctrl);
    this->burstStartTime = micros();
    this->pulseOn();
}

void BurstingState :: exit(StateController* ctrl) {
    this->pulseOff();
}

void BurstingState :: update(StateController* ctrl) {
    if (micros() - this->burstStartTime > this->timeout) {         // timeout reached, return to idle
        ctrl->go_to_state(IdleState::getInstance()); 
    }
}

void BurstingState :: set_parameters(StateController* ctrl) {
    this->timeout = ctrl->settings->timeout * 1e6;            // time in us
   
    this->ctrlPin = ctrl->settings->cs_burst_control;
    this->burstFreq = 1000.0f / (float) ctrl->settings->burst_period;
    this->dutyCycle = 100.0f - (float) ctrl->settings->duty_cycle;  // Invert because ctrlPin is active LOW
    
    if (!this->burst_ctrl) {
        this->burst_ctrl = new nRF52_PWM(this->ctrlPin, this->burstFreq, 100.0f);
    }
}

void BurstingState :: pulseOn(void) {
    this->burst_ctrl->setPWM(this->ctrlPin, this->burstFreq, this->dutyCycle);
}

void BurstingState :: pulseOff(void) {
    this->burst_ctrl->setPWM_manual(this->ctrlPin, 100.0f);
}

unsigned long BurstingState :: burst_elapsed(void) {
    return (micros() - this->burstStartTime);
}

File: lib/wus_control/bursting_state.py
import time
import board
from wus_control.state import State

class BurstingState(State):
    
    def __init__(self):
        self._ctrl_pin = None
        self._counter = None
        self._duty_cycle = 50

    @property
    def name(self):
        return 'bursting'

    def enter(self, controller):
        controller.rgb_led[0] = (0, 0, 255)
        State.enter(self, controller)
        self._set_parameters(controller)
        self._start_burst = time.monotonic()
        self._start_pulse = self._start_burst
        print('Bursting!')

    def exit(self, controller):
        State.exit(self, controller)
        self._off()

    def update(self, controller):
        status = State.update(self, controller)
        if status == 'NO_CONNECT':
            controller.go_to_state('advertising')
        elif status == 'INTERRUPT':
            controller.go_to_state('interrupt')
        elif self._burst_elapsed() > self._timeout:
            print('\tTimeout')
            controller.go_to_state('idle')
        else:
            if self._pulse_elapsed() > self._period:
                self._start_pulse = time.monotonic()
                self._on()
            elif self._pulse_elapsed() > self._on_time:
                self._off()

    def _set_parameters(self, controller):
        self._ctrl_pin = controller.settings['cs_burst_control']
        self._off()
        self._timeout = controller.settings['timeout']
        self._period = controller.settings['burst_period'] / 1000
        if not controller.settings['pulse_count']:
            self._duty_cycle = controller.settings['duty_cycle']
        self._on_time = self._period * self._duty_cycle

    def _on(self):
        self._ctrl_pin.value = False

    def _off(self):
        self._ctrl_pin.value = True

    def _burst_elapsed(self):
        return time.monotonic() - self._start_burst

    def _pulse_elapsed(self):
        return time.monotonic() - self._start_pulse
